core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -618,12 +655,9 @@ private void seekInternalWithCachingPartiallyReadBlock(long pos) throws IOExcept
       // The early return above guarantees that we won't close an incomplete cache stream.
       Preconditions.checkState(mCurrentCacheStream == null || mCurrentCacheStream.remaining() == 0);
       closeOrCancelCacheStream();
-    }
-
-    // If seeks within the current block, directly seeks to pos if we are not yet there.
-    // If seeks outside the current block, seek to the beginning of that block first, then
-    // cache the prefix (pos % mBlockSize) of that block.
-    if (isInCurrentBlock) {
+    } else if (isInCurrentBlock) {
+      // the current block is not cached, and the seek is within the block
```

Comment:
How do we know if the block is cached or not, or if we are reading from a remote worker?

Reply:
This is in the `else if` and the `if` branch checks if it needs to perform partial caching

Reply:
That makes sense.

I suggest to restructure the conditions, since currently we are combining two fairly unrelated blocks of logic.

From
```
if (...) // handle caching of the first block
else if (...) // handle pointer movement in some cases
// handle pointer movement in other cases and caching of the second block
```
to
```
if (...) // handle caching of the first block
if (...) // handle movement of pointer
if (...) // handle caching of the second block if necessary
```

Reply:
I don't think the structure is that easy to be rewritten to what you proposed.

The structure is like:
```
if (...) // handle caching of the current block
else if (...) handle pointer movement in the same block, and return early

if (...)  // handle caching of the target block
else (...) // handle none caching of the target block
```
You can see at high-level there are two main blocks: current block handling and target block handling
And within each section, there are conditions to handle different cases.

Reply:
How about this kind of condition structure?
```
if (...) // Current block does not need to be cached
  if (...) // Seek is in current block, move pointer
  else (...) // Move pointer to start of second block
else (...) // Current block needs to be cached
  (...) read block to some position
// Invariant after this section is the first block caching has been handled and the position is either at the correct position or the start of the second block
if (...) // position is at the start of the second block
  updateStreams()
  if (...) // need to partial cache second block
  else (...) // move the pointer to the correct position
```

Reply:
This structure is almost right.

In the else of `Current block needs to be cached`, you also need to move pointer to the start of the second block if it's in the same block.